packages/realm-web/src/WatchStream.ts

```diff
@@ -0,0 +1,270 @@
+     */
+    NEED_DATA = "NEED_DATA",
+    /**
+     * Call nextEvent() to consume an event.
+     */
+    HAVE_EVENT = "HAVE_EVENT",
+    /**
+     * Call error().
+     */
+    HAVE_ERROR = "HAVE_ERROR",
+}
+
+/**
+ * Represents a stream of events
+ */
+export class WatchStream<T> {
+    // Call these when you have data, in whatever shape is easiest for your SDK to get.
+    // Pick one, mixing and matching on a single instance isn't supported.
+    // These can only be called in NEED_DATA state, which is the initial state.
+    feedBuffer(buffer: Uint8Array) {
+        this.assertState(WatchStreamState.NEED_DATA);
+        this._buffer += this._textDecoder.decode(buffer, { stream: true });
+        this.advanceBufferState();
+    }
+
+    feedLine(line: string) {
+        this.assertState(WatchStreamState.NEED_DATA);
+        // This is an implementation of the algorithm described at
+        // https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation.
+        // Currently the server does not use id or retry lines, so that processing isn't implemented.
+
+        // ignore trailing LF if not removed by SDK.
+        if (line.endsWith("\n")) line = line.substr(0, line.length - 1);
+
+        // ignore trailing CR from CRLF
+        if (line.endsWith("\r")) line = line.substr(0, line.length - 1);
+
+        if (line.length == 0) {
+            // This is the "dispatch the event" portion of the algorithm.
+            if (this._dataBuffer.length == 0) {
+                this._eventType = "";
+                return;
+            }
+
+            if (this._dataBuffer.endsWith("\n"))
+                this._dataBuffer = this._dataBuffer.substr(
+                    0,
+                    this._dataBuffer.length - 1,
+                );
+
+            this.feedSse({
+                data: this._dataBuffer,
+                eventType: this._eventType,
+            });
+            this._dataBuffer = "";
+            this._eventType = "";
+        }
+
+        if (line[0] == ":") return;
+
+        const colon = line.indexOf(":");
+        const field = line.substr(0, colon);
+        let value = colon == -1 ? "" : line.substr(colon + 1);
+        if (value.startsWith(" ")) value = value.substr(1);
+
+        if (field == "event") {
+            this._eventType = value;
+        } else if (field == "data") {
+            this._dataBuffer += value;
+            this._dataBuffer += "\n";
+        } else {
+            // line is ignored (even if field is id or retry).
+        }
+    }
+
+    feedSse(sse: ServerSentEvent) {
+        this.assertState(WatchStreamState.NEED_DATA);
+        const firstPercentIndex = sse.data.indexOf("%");
+        if (firstPercentIndex != -1) {
+            // For some reason, the stich server decided to add percent-encoding for '%', '\n', and '\r' to its
+            // event-stream replies. But it isn't real urlencoding, since most characters pass through, so we can't use
```

Comment:
Did we notify the server team, this sounds like a partial double encoding...

Reply:
I havn't ... I agree that if there's a problem, we inform them to have it solved instead of just working around it.